Add EIP-5744: Latent Fungible Tokens (#5744)
ethereum/EIPs: EIPS/eip-5744.md

```solidity
struct MintMetadata {
  // Amount of tokens minted.
  uint256 amount;
  // Timestamp of the mint, in seconds.
  uint256 time;
  // Delay in seconds until these tokens mature and become fungible. When the
  // delay is not known (e.g. if it's dependent on other factors aside from
  // simply elapsed time), this value must be `type(uint256).max`.
  uint256 delay;
}

function getMints(address user) external view returns (MintMetadata[] memory);
```

Note that the implementation does not require that each of the above metadata parameters are stored as a `uint256`, just that they are returned as `uint256`.

An additional method called `mints` MAY be added.
This method returns the metadata for a mint based on its ID:

```solidity
function mints(address user, uint256 id) external view returns (MintMetadata memory);
```

The ID is not prescriptive—it may be an index in an array, or may be generated by other means.

The `transfer` and `transferFrom` methods MAY be modified to revert when transferring tokens that have not matured.
Similarly, any methods that burn tokens MAY be modified to revert when burning tokens that have not matured.

> All EIP-5744 tokens MUST implement EIP-20’s optional metadata extensions.

Reviewer: ```suggestion
All latent fungible tokens MUST implement EIP-20’s optional metadata extensions.
```

The `name` and `symbol` functions MUST reflect the underlying token’s `name` and `symbol` in some way.

## Rationale

The `mints` method is optional because the ID is optional. In some use cases such as vesting where a user may have a maximum of one mint, an ID is not required.

Similarly, vesting use cases may want to enforce non-transferrable tokens until maturity, whereas lending receipt tokens with a minimum deposit duration may want to support transfers at all times.

It is possible that the number of mints held by a user is so large that it is impractical to return all of them in a single `eth_call`.
This is unlikely so it was not included in the spec.
If this is likely for a given use case, the implementer may choose to implement an alternative method that returns a subset of the mints, such as `getMints(address user, uint256 startId, uint256 endId)`.
However, if IDs are not sequential, a different signature may be required, and therefore this was not included in the specification.

## Backwards Compatibility

> EIP-5744 is fully backward compatible with the EIP-20 standard and has no known compatibility issues with other standards.

Reviewer: ```suggestion
This proposal is fully backward compatible with the EIP-20 standard and has no known compatibility issues with other standards.
```

---
eip: 5744
title: Latent Fungible Token
description: An interface for tokens that become fungible after a period of time.
author: Cozy Finance (@cozyfinance), Tony Sheng (@tonysheng), Matt Solomon (@mds1), David Laprade (@davidlaprade), Payom Dousti (@payomdousti), Chad Fleming (@chad-js), Franz Chen (@Dendrimer)
discussions-to: https://ethereum-magicians.org/t/eip-5744-latent-fungible-token/11111
status: Draft
type: Standards Track
category: ERC
created: 2022-09-29
requires: 20, 2612
---

## Abstract

The following standard is an extension of [EIP-20](./eip-20.md) that enables tokens to become fungible after some initial non-fungible period.
Once minted, tokens are non-fungible until they reach maturity.

> At maturity, they become fungible and can be transferred, traded, and used in any way that a standard EIP-20 token can be used.

Author: IIUC you are saying confusion may stem from defining tokens as initially "non-fungible", and the worry is that this may be conflated with EIP-721, which is unrelated aside from the fact that they both use the adjective "non-fungible"?

If so, how you recommend rephrasing this? Perhaps adding a sentence explaining that EIP-721 is for tokens that are permanently non-fungible, and is therefore unrelated?

> At maturity, they become fungible and can be transferred, traded, and used in any way that a standard EIP-20 token can be used.

Reviewer: Yep, that's exactly my concern. I think calling these "latent fungible tokens" implies that up until the unlock time, the tokens are non-fungible instead of non-transferable.

## Motivation

Example use cases include:

- Receipt tokens that do not become active until a certain date or condition is met. For example, this can be used to enforce minimum deposit durations in lending protocols.
- Vesting tokens that cannot be transferred or used until the vesting period has elapsed.

## Specification

> All [EIP-5744](./eip-5744.md) tokens MUST implement EIP-20 to represent the token.

Author: Updated in 1f224c04

The `balanceOf` and `totalSupply` return quantities for all tokens, not just the matured, fungible tokens.
A new method called `balanceOfMatured` MUST be added to the ABI.
This method returns the balance of matured tokens for a given address:

```solidity
function balanceOfMatured(address user) external view returns (uint256);
```

An additional method called `getMints` MUST be added, which returns an array of all mint metadata for a given address:

```solidity
struct MintMetadata {
  // Amount of tokens minted.
  uint256 amount;
  // Timestamp of the mint, in seconds.
  uint256 time;
  // Delay in seconds until these tokens mature and become fungible. When the
  // delay is not known (e.g. if it's dependent on other factors aside from
  // simply elapsed time), this value must be `type(uint256).max`.
  uint256 delay;
}

function getMints(address user) external view returns (MintMetadata[] memory);
```

Note that the implementation does not require that each of the above metadata parameters are stored as a `uint256`, just that they are returned as `uint256`.

An additional method called `mints` MAY be added.

> This method returns the metadata for a mint based on it's ID:

Author: Fixed in 1f224c04

```solidity
function mints(address user, uint256 id) external view returns (MintMetadata memory);
```

The ID is not prescriptive—it may be an index in an array, or may be generated by other means.

The `transfer` and `transferFrom` methods MAY be modified to revert when transferring tokens that have not matured.

> Similarly, any methods that burn tokens MAY be modified to revert when burning tokens that have not matured.

Reviewer: So the locks are advisory only? It's possible to have an EIP-5744 compliant contract that allows you to transfer immature tokens?

Author: The thinking was that you _can_ implement all the logic to transfer immature tokens, but sometimes it will be complex/costly and other times it may be simple, so we didn't want to be too prescriptive here.

Two examples:
- A 4626-like vault that uses 5744 tokens as the receipt tokens: A user may have dozens or hundreds of mints, so allowing transfers of immature tokens can get costly
- A vesting use case where there's only 1 mint per user is much simpler, so allowing transfers prior to maturity seems reasonable (redemptions are still prevented until maturity)

Reviewer: Oooohhhhh, I think I missed the point on this entirely. I'm going to ask some questions in your discussion thread.

All EIP-5744 tokens MUST implement EIP-20’s optional metadata extensions.
The `name` and `symbol` functions MUST reflect the underlying token’s `name` and `symbol` in some way.

> EIP-5744 tokens MAY implement [EIP-2612](./eip-2612.md) to improve the UX of approving shares on various integrations.

Author: Not intending to prohibit that, I just took this line straight from [EIP-4626](https://eips.ethereum.org/EIPS/eip-4626#specification). So I removed it in 1f224c04

A new method called `balanceOfMatured` MUST be added to the ABI.
This method returns the balance of matured tokens for a given address:

```solidity
function balanceOfMatured(address user) external view returns (uint256);
```

An additional method called `getMints` MUST be added, which returns an array of all mint metadata for a given address:

```solidity
struct MintMetadata {
  // Amount of tokens minted.
  uint256 amount;
  // Timestamp of the mint, in seconds.
  uint256 time;
  // Delay in seconds until these tokens mature and become fungible. When the
  // delay is not known (e.g. if it's dependent on other factors aside from
  // simply elapsed time), this value must be `type(uint256).max`.
  uint256 delay;
}

function getMints(address user) external view returns (MintMetadata[] memory);
```

Note that the implementation does not require that each of the above metadata parameters are stored as a `uint256`, just that they are returned as `uint256`.

An additional method called `mints` MAY be added.
This method returns the metadata for a mint based on it's ID:

```solidity
function mints(address user, uint256 id) external view returns (MintMetadata memory);
```

> The ID is not prescriptive&mdash;it may be an index in an array, or may be generated by other means.

Reviewer 2: Please use the mdash unicode character.

Author: Done in b9b4f60f. Also committed your other suggestions from the GitHub UI which automatically marked those comments resolved

The `transfer` and `transferFrom` methods MAY be modified to revert when transferring tokens that have not matured.
Similarly, any methods that burn tokens MAY be modified to revert when burning tokens that have not matured.

All EIP-5744 tokens MUST implement EIP-20’s optional metadata extensions.

> The `name` and `symbol` functions SHOULD reflect the underlying token’s `name` and `symbol` in some way.

Reviewer 2: Why not must?

Author: I just used the same sentence here that EIP-4626 has in their specification. I'm ok with both SHOULD or MUST, let me know which you prefer

https://eips.ethereum.org/EIPS/eip-4626#specification

Reviewer 2: Personally, I prefer must. I see no use case where `name` and `symbol` shouldn't reflect the underlying token.

Author: I agree, I don't know why you'd want that 😅

Updated to MUST in 2361fad1

EIP-5744 tokens MAY implement [EIP-2612](./eip-2612.md) to improve the UX of approving shares on various integrations.

## Rationale

> The key design decisions here are the `balanceOfMatured` method and the `getMints` method.
> Other aspects may vary depending on the specific implementation.

Reviewer 2: ```suggestion
```

> For example, the `mints` method is optional because the ID is optional. Some use cases such as vesting where a user may have a maximum of one mint so an ID is not required.

Reviewer 2: ```suggestion
The `mints` method is optional because the ID is optional. In some use cases such as vesting where a user may have a maximum of one mint, an ID is not required.
```
